### PR TITLE
adds commands for listing deals

### PR DESCRIPTION
**Pull request by brunocalza**
Adds new commands for listing deals

```
basin publication deals —name PUBLICATION —latest N
basin publication deals —name PUBLICATION —limit N —offset EPOCH
```

Example of a real call:
```bash
publication deals --publication boulder.orca --latest 1 --provider localhost:3000

+----------+--------------------------------------------+-------------------------------------------------------------+
|    ID    |               SELECTOR PATH                |                             CID                             |
+----------+--------------------------------------------+-------------------------------------------------------------+
| 56194780 | Links/194/Hash/Links/165/Hash/Links/0/Hash | bafybeihxjbxiq2fbrlblcdp7o3r55fittn5vbsptqmfwivcvq77qqwzesa |
+----------+--------------------------------------------+-------------------------------------------------------------+
```

It is just returning the [DealInfo](https://github.com/tablelandnetwork/basin-storage/blob/main/evm/basin_storage/src/BasinStorage.sol#L14-L18) from the contract. Not sure how useful `ID` and `SELECTOR PATH` it is to the client.

### CODE REVIEW

File: internal/app/streamer.go [hunk 1]
```diff
@@ -22,14 +21,6 @@ type Replicator interface {
 	Shutdown()
 }
 
-// BasinProvider ...
```
Comment (brunocalza) [inline internal/app/streamer.go:25]: moved to a new file

File: cmd/basin/publication.go [hunk 2]
```diff
@@ -40,6 +41,7 @@ func newPublicationCommand() *cli.Command {
 			newPublicationStartCommand(),
 			newPublicationUploadCommand(),
 			newPublicationListCommand(),
+			newPublicationDealsCommand(),
```
Comment (brunocalza) [inline cmd/basin/publication.go:44]: adds new command for fetching deals

File: cmd/basin/publication.go [hunk 3]
```diff
@@ -330,6 +332,92 @@ func newPublicationListCommand() *cli.Command {
...
+			}
+
+			bp, err := basinprovider.New(cCtx.Context, provider)
+			if err != nil {
+				return fmt.Errorf("new basin provider: %s", err)
+			}
+			defer bp.Close()
+
+			var deals []app.DealInfo
+			if latest > 0 {
```
Comment (brunocalza) [inline cmd/basin/publication.go:388]: depending if latest is set or not calls one method `LatestDeals` or the other `Deals`

File: cmd/basin/publication.go [hunk 4]
```diff
@@ -330,6 +332,92 @@ func newPublicationListCommand() *cli.Command {
...
+				}
+			}
+
+			table := tablewriter.NewWriter(os.Stdout)
+			table.SetHeader([]string{"ID", "Selector Path", "CID"})
+
+			for _, deal := range deals {
+				table.Append([]string{fmt.Sprint(deal.ID), deal.SelectorPath, deal.CID})
+			}
+			table.Render()
```
Comment (avichalp) [inline cmd/basin/publication.go:414]: table view looks cool!

File: pkg/basinprovider/provider.go [hunk 5]
```diff
@@ -385,7 +492,85 @@ func (s *BasinServerMock) Upload(_ context.Context, call Publications_upload) er
...
+	if !ok {
+		s.uploads[pub] = make(map[int]*callbackMock)
+	}
+	s.uploads[pub][len(s.uploads[pub])] = callback
+
+	return nil
+}
+
+// Deals lists deals from a given publication.
+func (s *BasinServerMock) Deals(_ context.Context, call Publications_deals) error {
```
Comment (brunocalza) [inline pkg/basinprovider/provider.go:506]: this mock implementation is just for testing. it kind of simulates deal uploads

**Pull request merged**